apps/client/src/features/modals/sheets-modal/SheetsModal.tsx

```diff
@@ -0,0 +1,250 @@
+import { ChangeEvent, useRef, useState } from 'react';
+import {
+  Button,
+  Input,
+  Modal,
+  ModalBody,
+  ModalCloseButton,
+  ModalContent,
+  ModalFooter,
+  ModalHeader,
+  ModalOverlay,
+} from '@chakra-ui/react';
+import { IoCheckmarkCircleOutline } from '@react-icons/all-files/io5/IoCheckmarkCircleOutline';
+import { IoCloseCircleOutline } from '@react-icons/all-files/io5/IoCloseCircleOutline';
+import { useQueryClient } from '@tanstack/react-query';
+import { GoogleSheetState, OntimeRundown, ProjectData, UserFields } from 'ontime-types';
+
+import { PROJECT_DATA, RUNDOWN, USERFIELDS } from '../../../common/api/apiConstants';
+import { maybeAxiosError } from '../../../common/api/apiUtils';
+import {
+  getSheetsAuthUrl,
+  getSheetSettings,
+  getSheetstate,
+  patchData,
+  postPreviewSheet,
+  postPushSheet,
+  postSheetSettings,
+  uploadSheetClientFile,
+} from '../../../common/api/ontimeApi';
+import { projectDataPlaceholder } from '../../../common/models/ProjectData';
+import { userFieldsPlaceholder } from '../../../common/models/UserFields';
+import PreviewExcel from '../upload-modal/preview/PreviewExcel';
+
+interface SheetsModalProps {
+  onClose: () => void;
+  isOpen: boolean;
+}
+
+export default function SheetsModal(props: SheetsModalProps) {
+  const { isOpen, onClose } = props;
+
+  const fileInputRef = useRef<HTMLInputElement>(null);
+
+  const [rundown, setRundown] = useState<OntimeRundown | null>(null);
+  const [userFields, setUserFields] = useState<UserFields | null>(null);
+  const [project, setProject] = useState<ProjectData | null>(null);
+
+  const [sheetState, setSheetState] = useState<GoogleSheetState>({ auth: false, id: false, worksheet: false });
+
+  const queryClient = useQueryClient();
+
+  const sheetid = useRef<HTMLInputElement>(null);
+  const worksheet = useRef<HTMLInputElement>(null);
+
+  const handleClose = () => {
+    setRundown(null);
+    setProject(null);
+    setUserFields(null);
+    onClose();
+  };
+  const handleClick = () => {
+    fileInputRef.current?.click();
+  };
+
+  const handleFile = async (event: ChangeEvent<HTMLInputElement>) => {
+    const selectedFile = event?.target?.files?.[0];
+    if (!selectedFile) {
+      return;
+    } else {
+      await uploadSheetClientFile(selectedFile);
+      _onChange();
+    }
+  };
+
+  const _onChange = async () => {
+    setSheetState(await getSheetstate());
+  };
+
+  if (isOpen) {
+    //TODO: how to get this on modal open
```

Comment:
@cpvalente could you give a tip on this
I would like it to run this onOpen

Reply:
To run side effects in react, you would generally use a `useEffect`
Keep in mind that is not always the case, the recipe for useEffect is to synchronise react with things outside of react


In your case,
It seems that `getSheetState` should be called in a `useEffect` and its value pushed to a piece of state instead of a ref. You need the state to trigger a re-render in the components.

We can consider whether we want to push this into react-query which would manage this lifecycle for us. see how the others are done such as `useViewSettings`

Send me a message with what you are trying to achieve and I can take a look